msgpack/src/main/scala/fs2/data/msgpack/low/internal/ItemValidator.scala

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2024 fs2-data Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fs2
+package data
+package msgpack
+package low
+package internal
+
+case class ValidationErrorAt(at: Long, msg: String) extends Error(s"at position ${at}: ${msg}")
+case class ValidationError(msg: String) extends Exception(msg)
+
+private[low] object ItemValidator {
+
+  case class Expect(n: Int, from: Long) {
+    def dec = Expect(n - 1, from)
+  }
+
+  type ValidationContext = (Chunk[MsgpackItem], Int, Long, List[Expect])
+
+  def pipe[F[_]](implicit F: RaiseThrowable[F]): Pipe[F, MsgpackItem, MsgpackItem] = { in =>
+    def step1(chunk: Chunk[MsgpackItem], idx: Int, position: Long): Pull[F, MsgpackItem, Option[Expect]] =
+      chunk(idx) match {
+        case MsgpackItem.UnsignedInt(bytes) =>
+          if (bytes.size > 8) Pull.raiseError(new ValidationErrorAt(position, "Unsigned int exceeds 64 bits"))
+          else Pull.pure(None)
```

Comment:
Since we expect to have many of them, I would create a constant like `private final val PullNone = Pull.pure(None)`. This would avoid a lot of allocations.

Reply:
What is the difference between having `final val` and `val` in an object/method? I know that this keyword matters when defining a class field but I'm not aware of it's uses in other contextxs

Reply:
In this case it does not matter, generated bytecode will be the same. It's just some sort of muscle memory from my java days that came back to life :laughing: